notebooks/platform/xarray/lib/io/pybgen_backend.py

```diff
@@ -0,0 +1,91 @@
+from pybgen import PyBGEN
+from pybgen.parallel import ParallelPyBGEN
+import xarray as xr
+
+from .. import core
+from ..typing import PathType
+from ..compat import Requirement
+from ..dispatch import ClassBackend, register_backend
+from .core import BGEN_DOMAIN
+
+
+def _array_name(f, path):
+    return f.__qualname__ + ':' + str(path)
+
+
+class BgenReader(object):
+
+    def __init__(self, path, dtype=np.float32):
+        self.path = str(path) # pybgen needs string paths
+
+        # Use ParallelPyBGEN only to get all the variant seek positions from the BGEN index.
+        # No parallel IO happens here.
+        with ParallelPyBGEN(self.path, probs_only=False) as bgen:
+            bgen._get_all_seeks()
+            self._seeks = bgen._seeks
+            n_variants = bgen.nb_variants
+            n_samples = bgen.nb_samples
+
+            self.shape = (n_variants, n_samples)
+            self.dtype = dtype
+            self.ndim = 2
+
+            self.sample_id = bgen.samples
+            # This may need chunking for large numbers of variants
+            variants = list(bgen.iter_variant_info())
+            self.variant_id = [v.name for v in variants]
+            self.contig = [v.chrom for v in variants]
+            self.pos = [v.pos for v in variants]
+            self.a1 = [v.a1 for v in variants]
+            self.a2 = [v.a2 for v in variants]
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, tuple):
+            raise IndexError(f'Indexer must be tuple (received {type(idx)})')
+        if len(idx) != self.ndim:
+            raise IndexError(f'Indexer must be two-item tuple (received {len(idx)} slices)')
+
+        # Restrict to seeks for this chunk
+        seeks_for_chunk = self._seeks[idx[0]]
+        if len(seeks_for_chunk) == 0:
+            return np.empty((0, 0), dtype=self.dtype)
+        with PyBGEN(self.path, probs_only=False) as bgen:
+            p = [probs for (_, probs) in bgen._iter_seeks(seeks_for_chunk)]
+            return np.stack(p)[:,idx[1]]
+
+
+@register_backend(BGEN_DOMAIN)
+class PyBgenBackend(ClassBackend):
+
+    id = 'pybgen'
+
+    def read_bgen(self, path: PathType, chunks='auto', lock=False):
+
+        bgen_reader = BgenReader(path)
+
+        vars = {
+            "sample_id": xr.DataArray(np.array(bgen_reader.sample_id), dims=["sample"]),
+            "variant_id": xr.DataArray(np.array(bgen_reader.variant_id), dims=["variant"]),
+            "contig": xr.DataArray(np.array(bgen_reader.contig), dims=["variant"]),
+            "pos": xr.DataArray(np.array(bgen_reader.pos), dims=["variant"]),
+            "a1": xr.DataArray(np.array(bgen_reader.a1), dims=["variant"]),
+            "a2": xr.DataArray(np.array(bgen_reader.a2), dims=["variant"]),
+        }
+
+        arr = da.from_array(
+            bgen_reader,
+            chunks=chunks,
+            lock=lock,
+            asarray=False,
+            name=_array_name(self.read_bgen, path))
+
+        # pylint: disable=no-member
```

Comment:
There's a short history to the need for this that you might immediately see a good way to fix.  All of the dataset classes in [core.py](https://github.com/related-sciences/gwas-analysis/blob/master/notebooks/platform/xarray/lib/core.py) are not true subclasses of an Xarray dataset, which was a choice made based at least in part on https://github.com/pydata/xarray/issues/3959.  Because of that, the solution I went with on these constructor methods was to monkey patch them in based on the class names.  I only wanted them there because `GenotypeDosageDataset.create` seemed less idiomatic to me than a bunch of `create_*` functions.  Do you know how to accomplish something like that without monkey-patching?

Reply:
I didn't look into this much at the time, but I wonder why there isn't just a regular `create_genotype_dosage_dataset` function (plus the others)?

Maybe there's no need to have classes here at all - since the only role is that of a tag being stored in an xarray attribute. You'd still need that tag, but the `to_` methods could just be functions (or methods on xarray like they are now), that just throw an error if that conversion isn't supported. Not sure if that's a better way to do things.

Reply:
My rationale for using the classes was mainly that they were a natural place to hang things like names and numbers of dimensions expected, properties that triggered value-based preconditions (e.g. probabilities are in [0, 1] and sum to 1 along an axis, counts are non-negative, dosages don't exceed ploidy, etc.), and the conversion methods to avoid awkwardly-named functions like `call_dataset_to_dosage_dataset` for every valid combination.  One thing I know for sure is that of all the pairs of dataset types, more combinations of conversions are NOT supported than are supported, so I wanted to somehow make the valid combos clear.  

I waffled a bit between doing that all with a functions-only API in the beginning, so I'm happy to commit in that direction if you think it's worth it.

On a related topic, I'd love to be able to put expensive preconditions like those I mentioned into Dask task graphs rather than forcing them to run eagerly.  It would also be nice to make them optional for users that have some more confidence in the data on the way in.  I imagine you've seen this many times in your career -- do you think there is a good model in another project we ought to follow for doing these kinds of checks in the least intrusive way possible?  I don't see things like this often in other domain-specific PyData projects but I'm not so sure we should follow that lead.

Reply:
Thanks for the explanation. Given that most combinations *don't* exists then it makes sense to do something like you have now, or a more functional style. Perhaps it would help to spell out which combinations can be implemented, and which are just impossible.

Precondition checks are not things I've seen much of to be honest. That's probably because most things fail fairly fast, and not after a long computation. In this case it sounds like the computation might not fail, just give bogus results since the input was malformed in some way. How about writing some validation functions that are standalone, and if they turn out to be useful they could be combined with the main functions in some way.

Reply:
> How about writing some validation functions that are standalone, and if they turn out to be useful they could be combined with the main functions in some way

Sounds reasonable.  TF assertions are my mental model for this and since they don't exist in Dask, there probably isn't a good way to do it efficiently now.  Eager evaluation in a separate set of functions seems like a more natural fit, now that you mention it.  They'll probably whither more over time that way but maybe preventing bogus results is easier to add in down the road than I'm imagining (specifically when going from VCF to one of the more analysis-ready xarray dataset types I had scoped out).